tsfc/coffee_mode.py

```diff
@@ -0,0 +1,294 @@
+
+    ilp_prob.solve()
+    if ilp_prob.status != 1:
+        raise RuntimeError("Something bad happened during ILP")
+
+    def optimal(atomic):
+        return ilp_var[atomic_index[atomic]].value() == 1
+
+    optimal_atomics = filter(optimal, iterkeys(atomic_index))
+    return tuple(optimal_atomics)
+
+
+def factorise_atomics(monomial_sum, optimal_atomics, argument_indices):
+    """Group and factorise monomials using a list of atomics as common
+    subexpressions. Create new monomials for each group and optimise them recursively.
+
+    :arg monomial_sum: A :class:`MonomialSum` object, or a iterable collection
+                       of monomials, all of the monomials should have the
+                       same sum indices
+    :arg optimal_atomics: list of tuples of atomics to be used as common subexpression
+    :arg argument_indices: tuple of argument indices
+
+    :returns: a factorised :class:`MonomialSum` object, or the original object
+    if no changes are made
+    """
+    if not optimal_atomics or len(monomial_sum) < 2:
+        # Nothing to do
+        if isinstance(monomial_sum, MonomialSum):
+            return monomial_sum
+        else:
+            # input is list of monomials
+            new_monomial_sum = MonomialSum()
+            for m in monomial_sum:
+                new_monomial_sum.add(*m)
+            return new_monomial_sum
+    new_monomial_sum = MonomialSum()
+    # Group monomials with respect to each optimal atomic
+    factor_group = OrderedDict()
+    for monomial in monomial_sum:
+        for oa in optimal_atomics:
+            if oa in monomial.atomics:
+                # Add monomial to the list of corresponding optimal atomic
+                factor_group.setdefault(oa, []).append(monomial)
+                break
+        else:
+            # Add monomials that do no have argument factors to new MonomialSum
+            new_monomial_sum.add(*monomial)
+    # We should not drop monomials
+    assert sum(map(len, itervalues(factor_group))) + len(new_monomial_sum) == len(monomial_sum)
+
+    for oa, monomials in iteritems(factor_group):
+        if len(monomials) == 1:
+            # Just one monomial with this group, add to new MonomialSum straightaway
+            monomial, = monomials
+            new_monomial_sum.add(*monomial)
+            continue
+        all_atomics = []  # collect all atomics from monomials
+        all_rest = []  # collect all rest from monomials
+        for monomial in monomials:
+            _atomics = list(monomial.atomics)
+            _atomics.remove(oa)  # remove common factor
+            all_atomics.append(_atomics)
+            all_rest.append(monomial.rest)
+        # Create new MonomialSum for the factorised out terms
+        sub_monomial_sum = MonomialSum()
+        for _atomics, _rest in zip(all_atomics, all_rest):
+            sub_monomial_sum.add((), _atomics, _rest)
+        sub_monomial_sum = optimise_monomial_sum(sub_monomial_sum, argument_indices)
```

Comment:
This call does too much work, because it groups on sum_indices, but we know sub_monomial_sum only has sum_indices `()`.  So factor out core part and call that.